### ISSUE TITLE
Local function causes problems with an If statement

### ISSUE THREAD

**Issue by odytrice** [dotnet/roslyn#21324]
A local function in the below snippet of code causes the if statement to execute even though the condition is false. Removing the local function fixes the problem

**Version Used**: Not Sure but I have tried it with VS 2017 latest update and the compiler in LinqPad

**Steps to Reproduce**:

```csharp
void Main()
{
	//List of Claims
	List<Claim> claims = new List<Claim> { new Claim(ClaimTypes.NameIdentifier,"123") };

	Claim extClaims = null;
	if (extClaims == null)
	{
		extClaims = claims.FirstOrDefault(x => x.Type == ClaimTypes.GivenName);
	}
	if(extClaims == null)
	{
		extClaims = claims.FirstOrDefault(x => x.Type == ClaimTypes.NameIdentifier);
	}
	if (extClaims == null) throw new Exception("Unknown userId");
	
	void GetClaimValue() { }
}
```

**Expected Behavior**: Nothing

**Actual Behavior**: ~~Throws Exception~~ Debugger Enters `if` block and exits execution

**Comment by jskeet** [comment]
Are you able to convert this to a complete example? I've just tried it with simple `Claim` and `ClaimTypes` types, and can't reproduce...

**Comment by sharwell** [comment]
@odytrice How did you determine that an exception is thrown?

I ask because there were some bugs in the past related to sequence points that affected the debugger experience, making users think exceptions were thrown when in reality they were not.

Wrapping the `throw` statement in a block would fix the debugger issue:

```csharp
if (extClaims == null)
{
  throw new Exception("Unknown userId");
}
```

**Comment by odytrice** [comment]
I just noticed the problem. Its not exactly throwing the exception. But try stepping through it, It enters the `if` block but does nothing. This piece of code was in a web application and entering that block stops execution of the application and as such, I thought it was an exception

**Comment by sharwell** [comment]
@odytrice If that was the problem, let me know so I can go find the original and mark this as a duplicate. I just didn't want to do that before hearing from you that we really found the issue. 😄

**Comment by odytrice** [comment]
Yup, that was the problem

**Comment by sharwell** [comment]
@dotnet/roslyn-interactive I wasn't able to find the duplicate for this specific issue. Can you verify that these sequence points were fixed and indicate which bug/PR it was? Thanks!

**Comment by tmat** [comment]
@jcouv

**Comment by jcouv** [comment]
Tagging @agocke for local functions

**Comment by fdradmin** [comment]
Hi - same issue here...when debugging an if(false) statement followed immediately by a local function declaration, the debugger appears to step into the if(false) statement's body.  The code isn't executed, but even seeing the debugger step into the statement body is alarming and resulted in some wasted time before I realized what was going on.  I've created a simple repro, along with a 30 second MP4 that shows the VS debugger in action.  I'm using VS 2017 Enterprise 15.2 (26430.13).

Repro VS sln/ video:
[LocalFuncDebugWeirdness.zip](https://github.com/dotnet/roslyn/files/1327652/LocalFuncDebugWeirdness.zip)

Relevant Code:
````c#
static void Main(string[] args)
{
    int smaller = 2;
    int larger = 6;
    if (smaller > larger)
    {
        //Debugger will step into this block, but it isn't actually executed...
        return;
    }

     //Uncomment this line, and the issue goes away - debugger acts as expected
    //int x = DateTime.Now.Minute;

    //Local func definition...if we put any code between the if(false) statement above and here
    //(i.e. uncomment the "x =" line above), the problem is resolved; it seems to only occur if
    //a local func def immediately follows the if statement.
    string CreateString()
    {
        return "hello";
    }

    //Just here so you can see the debugger step past the function def
    Console.Write(1);
}
````

**Comment by jcouv** [comment]
Thanks for the details. I'll investigate in the coming week.
It does look like a sequence point issue (code executes fine, but debugger makes it seem like some un-executed code was reached). 
May relate to https://github.com/dotnet/roslyn/issues/17076 (another sequence point issue with `if` statements, that I've not yet been able to isolate)

**Comment by jcouv** [comment]
I think the problem is already fixed in 15.4 preview 3 at least.

When stepping through the [code sample](https://github.com/dotnet/roslyn/issues/21324#issuecomment-331733547) from the thread above, I see these two steps, which are correct. The conditional `return` is never executed or highlighted in debugger.

![image](https://user-images.githubusercontent.com/12466233/31143734-11d140f4-a833-11e7-87a4-7cd8c01c593d.png)

![image](https://user-images.githubusercontent.com/12466233/31143747-1c833412-a833-11e7-992a-698015ff28b0.png)

I'll go ahead and close the issue. Feel free to re-open if you still encounter this kind of problem in 15.4 or later. Thanks
